Feature: add TargetLost event

SlayerDF opened the pull request:
There is no way to know whether `AgentBehaviour`'s current target is missing using events. `TargetChanged` is not invoking in this case that's why I added this feature.

My use case: I have AgentMoveBehaviour to follow near enemies which are being detected using sensors. When the enemy goes out of the sensor radius, there is no way to prevent current movement using already implemented events.

crashkonijn commented:
Hi there!

Thanks for checking out my package and doing a PR! <3

You did indeed find a case where the event isn't thrown however there are more cases where the `OnTargetChanged` is already thrown with a null target.

If we remove [this line](https://github.com/crashkonijn/GOAP/blob/feature/v2.2/Package/Runtime/CrashKonijn.Agent.Runtime/AgentBehaviour.cs#L203) the `OnTargetChanged` event should also be thrown with a null value [here](https://github.com/crashkonijn/GOAP/blob/feature/v2.2/Package/Runtime/CrashKonijn.Agent.Runtime/AgentBehaviour.cs#L100).

Either that, or the `UpdateTarget` should probably also throw the new `TargetLost` event when the value is null. If this is added, should the `OnTargetChanges` event still be thrown when the value is null?

Let me know what you think!

SlayerDF commented:
Hi!

> If we remove [this line](https://github.com/crashkonijn/GOAP/blob/feature/v2.2/Package/Runtime/CrashKonijn.Agent.Runtime/AgentBehaviour.cs#L203) the OnTargetChanged event should also be thrown with a null value [here](https://github.com/crashkonijn/GOAP/blob/feature/v2.2/Package/Runtime/CrashKonijn.Agent.Runtime/AgentBehaviour.cs#L100).

It won't be thrown because `ActionState` is reset by this point and `Run` method won't call `UpdateTarget`.

I've updated the code. I've moved `TargetLost` invocation into the `UpdateTarget`. Also I've moved the `SetState` call from the `Run` because, in my opinion, it fits better in `ResetAction`. I think `OnTargetChanged` should not be thrown with null at all, because otherwise `inRange` parameter doesn't have a meaning.  It seems more logical to me to only throw the new event in this case.

crashkonijn commented:
Great changes, thanks!

Only thing I did was remove the else statement to match the coding style I use (I should probably write this down 😁)

crashkonijn commented:
Released as 3.0.10-beta!
